### PR TITLE
[FEAT]: Add explicit response evaluation scopes

### PR DESCRIPTION
**Pull request by spencrr**
## Description

Adds `ResponseScope.ANY_TURN`, `ALL_TURNS`, and `CURRENT_TURN` to `ResponseContains`. Omitting `scope` preserves current-response behavior and emits a `FutureWarning` for multi-turn contexts, prompting callers to make the intended quantifier explicit.

The documentation includes one source-of-truth table for positive and negated attack/probe forms, explains how each built-in evaluator sees the transcript, and clarifies that scope applies to turns already present in `EvalContext` rather than controlling execution length or stopping.

## Breaking changes

None. The API is additive and omitted scope retains existing behavior. Multi-turn calls without an explicit scope now emit a `FutureWarning`.

## Checklist

- [x] `pre-commit run --all-files` passes
- [x] Tests added for all response scopes, warnings, empty contexts, and negation
- [x] Documentation updated

Validation: 31 focused evaluator tests pass. Lint, formatting, type checks, and strict documentation build pass.

### CODE REVIEW

File: docs/attacks/xpia.md [hunk 1]
```diff
@@ -121,26 +121,44 @@ For generating many variants at scale, use [`Payloads.generate_async()`][rampart
...
-    & ResponseContains("id_rsa")
+    & ResponseContains("id_rsa", scope=ResponseScope.ANY_TURN)
 )
 
-# NOT: agent did NOT refuse — it complied with the injection
-evaluator = ~ResponseContains(lambda text: "I can't" in text or "I cannot" in text)
+# NOT: the agent failed to refuse on at least one turn
+evaluator = ~ResponseContains(
+    lambda text: "I can't" in text or "I cannot" in text,
+    scope=ResponseScope.ALL_TURNS,
```
Comment (nina-msft) [inline docs/attacks/xpia.md:144]: Why is this `ALL_TURNS` instead of `ANY_TURN`? Is this because we are negating?

What would happen if we did `~ResponseContains` for ANY_TURNS?

File: docs/attacks/xpia.md [hunk 2]
```diff
@@ -121,26 +121,44 @@ For generating many variants at scale, use [`Payloads.generate_async()`][rampart
...
 Place the cheaper evaluator on the left side of `|` — it short-circuits if the left operand detects.
 
+!!! warning "Multi-turn scope"
+    State the temporal scope explicitly for multi-turn attacks. Use
+    `ANY_TURN` for "leaked at some point" and negate `ALL_TURNS` for "failed
+    to refuse at least once." Omitting `scope` inspects only the current
+    response and emits a `FutureWarning` for multi-turn transcripts. See
+    [Temporal Scope](../usage/authoring-tests.md#temporal-scope).
+
+    This release prepares evaluator semantics for final-trace verdicts. Until
```
Comment (nina-msft) [inline docs/attacks/xpia.md]: doesn't impact this note but do we have a plan for when this change might ship?

File: docs/contributing/extending-rampart.md [hunk 3]
```diff
@@ -266,6 +268,15 @@ class MyEvaluator(BaseEvaluator):
 
 Evaluator tests should cover detection, non-detection, edge cases (empty response, missing data), and that `evidence` / `rationale` are populated correctly.
 
+!!! warning "Multi-turn evaluator migration"
+    Final-trace verdicts call an evaluator once with the complete transcript.
+    A custom evaluator that reads only `context.turns[-1]` intentionally judges
+    only the terminal response and cannot preserve earlier evidence. Rewrite
+    multi-turn predicates to inspect `context.turns` explicitly before
+    migrating execution cadence. The worked execution-strategy loop elsewhere
```
Comment (nina-msft) [inline docs/contributing/extending-rampart.md]: This note would be more clear if we called out explicitly where we mean by "elsewhere on this page" - what line number?

File: docs/probes/behavioral.md [hunk 4]
```diff
@@ -54,20 +54,47 @@ result = await Probes.behavior(
...
+These are the migration forms for complete-transcript probe requirements:
+
+```python
+from rampart.evaluators import ResponseContains, ResponseScope
+
+# Every response must contain the expected term
+ResponseContains("Paris", scope=ResponseScope.ALL_TURNS)
+
+# No response may contain the forbidden term
+~ResponseContains("password", scope=ResponseScope.ANY_TURN)
```
Comment (nina-msft) [inline docs/probes/behavioral.md]: In addition to these, should we have examples for `~ResponseContains(ALL_TURNS)` and `ResponseContains(ANY_TURN)` to make it clear what behavior in all cases is?

File: docs/probes/behavioral.md [hunk 5]
```diff
@@ -54,20 +54,47 @@ result = await Probes.behavior(
...
+
+!!! warning "Multi-turn scope"
+    Omitting `scope` inspects only the current response and emits a
+    `FutureWarning` for multi-turn transcripts. See
+    [Temporal Scope](../usage/authoring-tests.md#temporal-scope).
+
+    This release prepares evaluator semantics for final-trace verdicts. Probe
+    executions still stop on the first detected prefix, so `ALL_TURNS` and
+    negated `ANY_TURN` cannot yet enforce requirements on prompts that were
+    never sent. Choose an explicit scope now, but rely on the complete
```
Comment (nina-msft) [inline docs/probes/behavioral.md]: I don't love the notes that say to wait for specific things to land because when would users know that specific change landed? Maybe we just stick to guidance for today, and then update the note as needed when behavior changes

File: docs/usage/authoring-tests.md [hunk 6]
```diff
@@ -117,6 +117,45 @@ ResponseContains(re.compile(r"ssh-rsa\s+[A-Za-z0-9+/]+"))
 ResponseContains(lambda text: "secret" in text.lower())
 ```
 
+#### Temporal Scope
+
+By default, `ResponseContains` inspects only the current response. For a
+multi-turn transcript, pass an explicit
+[`ResponseScope`][rampart.evaluators.response_contains.ResponseScope]:
```
Comment (nina-msft) [inline docs/usage/authoring-tests.md:128]: nit: is this supposed to be a link? right now the reference is just noted after `ResponseScope`

File: docs/usage/authoring-tests.md [hunk 7]
```diff
@@ -117,6 +117,45 @@ ResponseContains(re.compile(r"ssh-rsa\s+[A-Za-z0-9+/]+"))
...
+
+# Detect only if every response contained the pattern
+ResponseContains("Paris", scope=ResponseScope.ALL_TURNS)
+
+# Inspect only the latest response and ignore earlier turns
+ResponseContains("id_rsa", scope=ResponseScope.CURRENT_TURN)
+```
+
+| Existing use | Intended meaning | Explicit form |
+|---|---|---|
```
Comment (nina-msft) [inline docs/usage/authoring-tests.md:144]: Ok this table is a good reference - maybe in the other docs files we can reference this as source of truth to avoid confusion :D

File: rampart/evaluators/response_contains.py [hunk 8]
```diff
@@ -33,33 +58,88 @@ def __init__(
...
+        """Resolve the scope and warn about ambiguous multi-turn evaluation.
+
+        Returns:
+            ResponseScope: The configured scope, or CURRENT_TURN when omitted.
+        """
+        if self._scope is not None:
+            return self._scope
+        if len(context.turns) > 1:
+            warnings.warn(_UNSPECIFIED_SCOPE_WARNING, FutureWarning, stacklevel=3)
+        return ResponseScope.CURRENT_TURN
```
Comment (nina-msft) [inline rampart/evaluators/response_contains.py:203]: This PR makes a mental model gap more visible...

Across the built-in evaluators there are now four different temporal behaviors, arrived at three different ways:

- `ToolCalled` / `SideEffectOccurred` — implicit ∃ over all turns, no knob
- `LLMJudge` — `TranscriptScope` controls how much transcript is *rendered into the prompt*; the model then returns one holistic verdict
- `ResponseContains` — `ResponseScope` controls how per-turn boolean matches are *quantified* (∃ / ∀ / last-only)

For `TranscriptScope` / `ResponseScope` -- both enums expose a `CURRENT_TURN` member resolving to the same `turns[-1]` window, so they read as interchangeable in a test file while being type-incompatible.

A short subsection in `docs/usage/authoring-tests.md` would close it — something like:

```markdown
#### How Each Evaluator Sees the Transcript

Evaluators reach their temporal behavior in two different ways. *Quantifying*
evaluators compute a per-turn result and combine them; *windowing* evaluators
choose how much transcript to hand to a judge that returns one verdict.

| Evaluator | Mechanism | Default | Configurable via |
|---|---|---|---|
| `ToolCalled` | quantifies (∃ over all turns) | any turn | — |
| `SideEffectOccurred` | quantifies (∃ over all turns) | any turn | — |
| `ResponseContains` | quantifies | current turn | `ResponseScope` |
| `LLMJudge` | windows | full transcript | `TranscriptScope` |

`ResponseScope.CURRENT_TURN` and `TranscriptScope.CURRENT_TURN` both narrow to
the last turn, but they belong to different enums and are not interchangeable —
pass the one your evaluator declares.
```

File: rampart/evaluators/response_contains.py [hunk 9]
```diff
@@ -15,6 +17,27 @@
...
+
+    Attributes:
+        ANY_TURN: Detect when at least one response matches.
+        ALL_TURNS: Detect only when every response matches.
+        CURRENT_TURN: Inspect only the most recent response.
+    """
+
+    ANY_TURN = "any_turn"
+    ALL_TURNS = "all_turns"
+    CURRENT_TURN = "current_turn"
```
Comment (nina-msft) [inline rampart/evaluators/response_contains.py:35]: The migration caveat only lives in the prose docs, but `docs/api/evaluators.md` renders this enum straight from the docstring — so anyone who finds `ResponseScope` via the API reference or editor autocomplete sees three equal-looking options with no hint that two of them don't yet behave as advertised.

Since probe/attack executions still evaluate growing prefixes and break on the first `DETECTED`, `ALL_TURNS` and `ANY_TURN` can't actually enforce a whole-transcript quantifier today. That's a big enough gotcha to belong in the docstring itself.

```python
class ResponseScope(Enum):
    """Which responses in the transcript the evaluator inspects.

    ``ANY_TURN`` and ``ALL_TURNS`` state intent for final-trace
    evaluation. Executions currently evaluate growing prefixes and stop
    on the first ``DETECTED`` outcome, so neither quantifier is enforced
    over prompts that were never sent. ``CURRENT_TURN`` is exact today.

    Attributes:
        ANY_TURN: Detect when at least one response matches. Not yet
            enforced across turns that the execution never reached.
        ALL_TURNS: Detect only when every response matches. Not yet
            enforced across turns that the execution never reached.
        CURRENT_TURN: Inspect only the most recent response.
    """

    ANY_TURN = "any_turn"
    ALL_TURNS = "all_turns"
    CURRENT_TURN = "current_turn"
```

File: rampart/evaluators/response_contains.py [hunk 10]
```diff
@@ -70,3 +150,15 @@ async def evaluate_async(self, *, context: EvalContext) -> EvalResult:
             outcome=EvalOutcome.NOT_DETECTED,
             rationale="Target pattern not found in response text",
         )
```
Comment (nina-msft) [inline rampart/evaluators/response_contains.py:221]: `evidence` now has three different shapes depending on which branch you land in:

- `CURRENT_TURN` detected → `["Pattern found in response text"]` (no turn number)
- `ANY_TURN`/`ALL_TURNS` detected → `["Pattern found on turn(s): 0, 2"]`
- `ANY_TURN` not detected → `[]`, but `ALL_TURNS` not detected → `["Pattern missing on turn(s): 1"]`

Anything downstream that reads `EvalResult.evidence` (reporting sinks, `_AllEvaluator` which concatenates evidence from both operands) now gets a format that varies by scope. Emitting the same shape from `_evaluate_current_turn` makes the contract uniform and strictly more informative — nothing currently asserts on the old string:

```python
    def _evaluate_current_turn(self, *, context: EvalContext) -> EvalResult:
        """Evaluate only the most recent response.

        Returns:
            EvalResult: The current-turn detection result.
        """
        current = context.current_turn
        if self._matches(current.response.text):
            return EvalResult(
                outcome=EvalOutcome.DETECTED,
                evidence=[f"Pattern found on turn(s): {current.turn_number}"],
                rationale="Response contains target pattern",
            )

        return EvalResult(
            outcome=EvalOutcome.NOT_DETECTED,
            rationale="Target pattern not found in response text",
        )
```

Going through `context.current_turn` instead of `context.text` also gives you the turn number without a second lookup.

**Review by nina-msft — COMMENTED** [review]
These comments are GHCP flagged & some of them I've tweaked wording slightly. Leaving these here for full coverage :-)

File: rampart/evaluators/__init__.py [hunk 11]
```diff
@@ -3,22 +3,24 @@
...
-from rampart.evaluators.response_contains import ResponseContains
+from rampart.evaluators.response_contains import ResponseContains, ResponseScope
 from rampart.evaluators.side_effect import SideEffectOccurred
 from rampart.evaluators.tool_called import ToolCalled
 
 __all__ = [
     "NEUTRAL_EVALUATOR",
     "LLMJudge",
     "ResponseContains",
+    "ResponseScope",
```
Comment (nina-msft) [inline rampart/evaluators/__init__.py:23]: Small asymmetry: `TranscriptScope` is re-exported from the top-level `rampart` package (`rampart/__init__.py` line 45 and `__all__`), but `ResponseScope` isn't. Two scope enums for two evaluators, only one reachable from `import rampart`.

```python
# rampart/__init__.py
from rampart.evaluators import LLMJudge, ResponseScope, TranscriptScope

__all__ = [
    ...
    "Response",
    "ResponseScope",
    "Result",
    ...
]
```

(`ResponseScope` sorts between `Response` and `Result`.)

File: rampart/evaluators/response_contains.py [hunk 12]
```diff
@@ -33,33 +58,88 @@ def __init__(
...
+                [f"Pattern missing on turn(s): {', '.join(missing_turns)}"]
+                if scope is ResponseScope.ALL_TURNS
+                else []
+            ),
+            rationale=(
+                "No response contains the target pattern"
+                if scope is ResponseScope.ANY_TURN
+                else "Not every response contains the target pattern"
+            ),
+        )
```
Comment (nina-msft) [inline rampart/evaluators/response_contains.py:171]: `evaluate_async` is ~45 lines now and carries the empty check, scope resolution, the current-turn delegation, the quantifier, two turn-number comprehensions, and four inline conditional expressions choosing rationale/evidence. It's noticeably harder to follow than the `_evaluate_current_turn` helper right below it.

Pulling the quantified path out mirrors the existing helper and gets each branch under the 20-line guidance in `coding-standards.instructions.md`:

```python
    async def evaluate_async(self, *, context: EvalContext) -> EvalResult:
        """Check response text for the target pattern within the scope.

        Returns:
            EvalResult: DETECTED when the configured scope is satisfied;
                NOT_DETECTED otherwise.

        Raises:
            ValueError: If the evaluation context has no turns.
        """
        if not context.turns:
            msg = "No turns in context."
            raise ValueError(msg)

        scope = self._resolve_scope(context=context)
        if scope is ResponseScope.CURRENT_TURN:
            return self._evaluate_current_turn(context=context)
        return self._evaluate_quantified(context=context, scope=scope)

    def _evaluate_quantified(
        self,
        *,
        context: EvalContext,
        scope: ResponseScope,
    ) -> EvalResult:
        """Apply the ANY/ALL quantifier across every response in the trace.

        Returns:
            EvalResult: DETECTED when the quantifier is satisfied.
        """
        matches = [self._matches(turn.response.text) for turn in context.turns]
        any_turn = scope is ResponseScope.ANY_TURN

        if any(matches) if any_turn else all(matches):
            return EvalResult(
                outcome=EvalOutcome.DETECTED,
                evidence=[self._turns_label(turns=context.turns, matches=matches, wanted=True)],
                rationale=(
                    "At least one response contains the target pattern"
                    if any_turn
                    else "Every response contains the target pattern"
                ),
            )

        return EvalResult(
            outcome=EvalOutcome.NOT_DETECTED,
            evidence=(
                []
                if any_turn
                else [self._turns_label(turns=context.turns, matches=matches, wanted=False)]
            ),
            rationale=(
                "No response contains the target pattern"
                if any_turn
                else "Not every response contains the target pattern"
            ),
        )

    @staticmethod
    def _turns_label(*, turns: list[Turn], matches: list[bool], wanted: bool) -> str:
        """Name the turns whose match state equals ``wanted``.

        Returns:
            str: An evidence line listing the relevant turn numbers.
        """
        numbers = [
            str(turn.turn_number)
            for turn, matched in zip(turns, matches, strict=True)
            if matched == wanted
        ]
        verb = "found" if wanted else "missing"
        return f"Pattern {verb} on turn(s): {', '.join(numbers)}"
```

`Turn` would need adding to the `TYPE_CHECKING` block. Combines cleanly with the evidence-shape comment below — `_evaluate_current_turn` can call `_turns_label` too.

File: docs/attacks/xpia.md (hunk 1 above)
Comment (spencrr) [reply docs/attacks/xpia.md:144]: not(ALL) -> one of the turns complies
not(ANY) -> none of the turns refuse

File: docs/attacks/xpia.md (hunk 2 above)
Comment (spencrr) [reply docs/attacks/xpia.md]: I removed rollout- and PR-specific wording and rewrote the guidance around current behavior: scope applies to turns already present in `EvalContext`; it does not control execution length or stopping.

File: docs/contributing/extending-rampart.md (hunk 3 above)
Comment (spencrr) [reply docs/contributing/extending-rampart.md]: Agreed. Replaced the vague reference with a direct link to the Attack walkthrough.

File: docs/probes/behavioral.md (hunk 4 above)
Comment (spencrr) [reply docs/probes/behavioral.md]: The Temporal Scope table now covers all four attack/probe and positive/negated combinations. The other guides link to that table rather than maintaining duplicate examples.

File: docs/probes/behavioral.md (hunk 5 above)
Comment (spencrr) [reply docs/probes/behavioral.md]: Removed the forward-looking language and rollout PR links. The docs now state durable scope semantics without requiring readers to know release history.

File: docs/usage/authoring-tests.md (hunk 6 above)
Comment (spencrr) [reply docs/usage/authoring-tests.md:128]: Yes, this is a mkdocstrings/autorefs link to the generated API entry.

File: docs/usage/authoring-tests.md (hunk 7 above)
Comment (spencrr) [reply docs/usage/authoring-tests.md:144]: Done. The attack and probe guides now link to the Temporal Scope table as the canonical mapping.

File: rampart/evaluators/response_contains.py (hunk 8 above)
Comment (spencrr) [reply rampart/evaluators/response_contains.py:203]: Added a "How Each Evaluator Sees the Transcript" subsection. It distinguishes quantifying evaluators from windowing evaluators and calls out that `ResponseScope.CURRENT_TURN` and `TranscriptScope.CURRENT_TURN` are different enum types.

File: rampart/evaluators/response_contains.py (hunk 10 above)
Comment (spencrr) [reply rampart/evaluators/response_contains.py:221]: I kept the existing `CURRENT_TURN` evidence string to avoid changing an established output contract in an otherwise additive PR. Quantified scopes include turn numbers because multiple turns may contribute. We can normalize all evidence shapes separately if we want that behavioral change.

File: rampart/evaluators/__init__.py (hunk 11 above)
Comment (spencrr) [reply rampart/evaluators/__init__.py:23]: I kept `ResponseScope` under `rampart.evaluators`, alongside `ResponseContains`. `ResponseContains` itself is not top-level; `TranscriptScope` follows the existing top-level `LLMJudge` export. This avoids expanding the top-level API asymmetrically.

File: rampart/evaluators/response_contains.py (hunk 9 above)
Comment (spencrr) [reply rampart/evaluators/response_contains.py:35]: Addressed in a timeless form. The docstring now says that scope applies only to turns present in `EvalContext` and does not control how many turns execution produces or whether it stops early.

File: rampart/evaluators/response_contains.py (hunk 12 above)
Comment (spencrr) [reply rampart/evaluators/response_contains.py:171]: Refactored. `evaluate_async` now delegates quantified behavior to focused `ANY_TURN` and `ALL_TURNS` helpers, with shared turn-label formatting. All 31 focused tests pass.

File: rampart/evaluators/response_contains.py [hunk 13]
```diff
@@ -70,3 +219,15 @@ async def evaluate_async(self, *, context: EvalContext) -> EvalResult:
...
+        """Return whether one response matches the configured target."""
+        if isinstance(self._target, re.Pattern):
+            return bool(self._target.search(text))
+        if isinstance(self._target, str):
+            check_text = text if self._case_sensitive else text.lower()
+            check_target = (
+                self._target if self._case_sensitive else self._target.lower()
+            )
+            return check_target in check_text
+        return self._target(text)
```
Comment (nina-msft) [inline rampart/evaluators/response_contains.py:233]: non-blocking copilot find that should be a small one-lined fix + regression test:

`_matches` is annotated `-> bool` but this branch returns the predicate's value uncoerced, and `_turns_label` compares with `is` (line 189). So a predicate returning a truthy non-`bool` — `lambda t: re.search("id_rsa", t)` — never matches `True` and the turn list comes out empty:

```
ResponseContains(lambda t: re.search("id_rsa", t), scope=ANY_TURN)
  → DETECTED, evidence=['Pattern found on turn(s): ']
```

Outcome is still correct; only the evidence degrades. `CURRENT_TURN` is unaffected, and this worked before the PR since the old code did a plain truthy check.

```python
return bool(self._target(text))
```

I tried it locally — fixes all the quantified cases, leaves `CURRENT_TURN` identical, evaluator suite and ruff/ty still clean. Probably worth a regression test with a `re.search` lambda since nothing currently covers a non-`bool` predicate.

Comment (spencrr) [reply rampart/evaluators/response_contains.py:233]: target is `target: str | re.Pattern[str] | Callable[[str], bool],`

if not `re.Pattern` or `str` then, it's a `(str) -> bool` or should be caught be typechecker on `__init__`

**Review by nina-msft — APPROVED** [review]
:shipit:
